src/v/cluster/self_test_frontend.cc

```diff
@@ -0,0 +1,305 @@
+  , _connections(connections) {}
+
+ss::future<> self_test_frontend::start() { return ss::now(); }
+
+ss::future<> self_test_frontend::stop() { co_await _gate.close(); }
+
+ss::future<self_test_frontend::err_or_id_t> self_test_frontend::start_test() {
+    auto gate_holder = _gate.hold();
+
+    auto leader_opt = _leaders.local().get_leader(model::controller_ntp);
+    if (!leader_opt.has_value() || leader_opt != _self) {
+        // Only cluster leader can initiaite start_test() rpc calls
+        co_return self_test_status_report::failure;
+    }
+    const auto st = co_await status();
+    if (st.is_out_of_sync()) {
+        // Result contains inconsistencies, cancel and report why
+        co_await stop_test();
+        co_return self_test_status_report::out_of_sync;
+    } else if (st.is_test_active()) {
+        // no inconsistences, but test active, hasn't finished
+        co_return self_test_status_report::running;
+    }
+
+    /// Invoke command to start test on all nodes, using the same test id
+    const auto id = uuid_t::create();
+    const auto results = co_await invoke_on_all_nodes(
+      [id](model::node_id, auto handle) { return handle->start_test(id); });
+    co_return id;
+}
+
+ss::future<self_test_status_report> self_test_frontend::stop_test() {
+    auto gate_holder = _gate.hold();
+
+    auto leader_opt = _leaders.local().get_leader(model::controller_ntp);
+    if (!leader_opt.has_value() || leader_opt != _self) {
+        // Only cluster leader can initiaite start_test() rpc calls
+        co_return self_test_status_report::failure;
+    }
+
+    const auto st = co_await status();
+    if (!st.is_out_of_sync() && !st.is_test_active()) {
+        // No test is running
+        co_return self_test_status_report::idle;
+    }
+
+    auto results = co_await invoke_on_all_nodes(
+      [](model::node_id, auto handle) { return handle->stop_test(); });
+
+    /// The state may contain cancelled status
+    _state.merge(global_test_state(std::move(results)));
+
+    /// Alert the user if cancellation didn't go as planned, i.e. for some
+    /// reason some nodes are still running jobs, or the system is just not in
+    /// sync.
+    if (_state.is_out_of_sync()) {
+        co_return self_test_status_report::out_of_sync;
+    } else if (_state.is_test_active()) {
+        co_return self_test_status_report::failure;
+    }
+    co_return self_test_status_report::idle;
+}
+
+ss::future<self_test_frontend::global_test_state> self_test_frontend::status() {
+    auto gate_holder = _gate.hold();
+
+    auto result = co_await invoke_on_all_nodes(
+      [this](model::node_id id, auto handle) {
+          return handle->query_status(!_state.has_completed(id));
+      });
+
+    // Condense response from all nodes to a known state about the system
+    _state.merge(global_test_state(std::move(result)));
+
+    if (!_state.is_test_active()) {
+        /// test has finished
+        if (_state.current_uuid() == uuid_t{}) {
```

Comment:
I don't fully understand the usage of a test id here. If all the tests are going to be identical, why not just have each local test result either be usable or not usable based on how recently the local test was run? Then we'd know whether a test result is valid or not solely based on whether it's stale, rather only being useful by exactly one global test.

Reply:
Right I understand the mapping of local to global, but I wonder why that's desirable vs having local results being usable by multiple global tests. I suppose if a user wanted to run multiple tests in succession, the results will be more reasonable if we guaranteed a fresh run every time. Also if there are plans to extend the tests with different parameters having test ID seems important.

Reply:
Taking this advice testId will have much less of an importance in the next update, however i'm leaving the notion of test ID in, maybe it might be useful for the frontend to know if the system was out of sync or something like that.